File: joggles/protocol.py
"""Funky Glasses+ BLE protocol.

Recovered from the vendor APK (com.pinkysinyeeho.funkyglassesplus):
  - command table  -> model/data/Agreement.java
  - GATT UUIDs     -> com/cdbwsoft/library/AppConfig.java
  - AES key        -> constant inside lib/arm64-v8a/libAES.so

Wire format. Every write is a 16-byte AES-128-ECB block:

    [len][opcode ASCII...][args...][zero padding to 16]

`len` counts the opcode bytes plus its arguments, not the padding. Opcodes are
uppercase ASCII, so a correct decryption is obvious on sight - which is exactly
how the key was recovered.
"""
from __future__ import annotations

from Crypto.Cipher import AES

# Advertised service the vendor app filters on.
SERVICE_UUID = "0000fee9-0000-1000-8000-00805f9b34fb"

# Command channel: 16-byte encrypted frames.
CHAR_COMMAND = "d44bc439-abfd-45a2-b575-925416129600"
# Device -> host notifications (STYPE replies land here).
CHAR_NOTIFY = "d44bc439-abfd-45a2-b575-925416129601"
# Bulk pixel upload. Same encryption, one column per block.
CHAR_BULK_A = "d44bc439-abfd-45a2-b575-92541612960a"
CHAR_BULK_B = "d44bc439-abfd-45a2-b575-92541612960b"

# Manufacturer-data signature the app matches on: ASCII "TR\x00:".
BROADCAST_SIGNATURE = bytes([0x54, 0x52, 0x00, 0x3A])

KEY = bytes.fromhex("34522a5b7a6e492c08090a9d8d2a23f8")

BLOCK = 16

# STYPE reply -> (rows, cols). The device reports its own panel size.
PANEL_SIZES = {
    536: (5, 36),
    1248: (12, 48),
    1456: (14, 56),
    1664: (16, 64),
}


def encrypt(block: bytes) -> bytes:
    if len(block) != BLOCK:
        raise ValueError(f"frame must be {BLOCK} bytes, got {len(block)}")
    return AES.new(KEY, AES.MODE_ECB).encrypt(block)


def decrypt(block: bytes) -> bytes:
    return AES.new(KEY, AES.MODE_ECB).decrypt(block)


def frame(opcode: str, *args: int) -> bytes:
    """Build a plaintext 16-byte frame. Encrypt before writing."""
    body = opcode.encode("ascii") + bytes(args)
    if len(body) > BLOCK - 1:
        raise ValueError(f"frame body too long: {len(body)}")
    return bytes([len(body)]) + body.ljust(BLOCK - 1, b"\x00")


def parse_type(plaintext: bytes) -> tuple[int, int] | None:
    """Decode a STYPE reply into (rows, cols)."""
    n = plaintext[0]
    body = plaintext[1:1 + n].decode("ascii", "replace")
    if not body.startswith("STYPE"):
        return None
    dims = body[5:]
    if "X" not in dims:
        return None
    rows, _, cols = dims.partition("X")
    try:
        return int(rows), int(cols)
    except ValueError:
        return None


# --- Command table, transcribed from Agreement.java -----------------------

def enter_diy() -> bytes:
    """Enter DIY drawing mode. Send before uploading pixels."""
    return frame("SMVEW", 1)


def enter_diy_alt() -> bytes:
    return frame("SMVEW", 3)


def exit_diy(save: bool = False) -> bytes:
    return frame("SMVEW", 2 if save else 0)


def query_type() -> bytes:
    """Ask the panel to report its dimensions. Reply arrives on CHAR_NOTIFY."""
    return frame("STYPE")


def brightness(level: int) -> bytes:
    return frame("LIGHT", level)


def leds(on: bool) -> bytes:
    return frame("LEDON") if on else frame("LEDOFF")


def flashlight(on: bool) -> bytes:
    return frame("LIGHTON") if on else frame("LIGHTOFF")


def speed(value: int) -> bytes:
    return frame("SPEED", value)


def invert() -> bytes:
    return frame("EVERT")


def animation(index: int) -> bytes:
    return frame("ANIM", index)


def animation_loop() -> bytes:
    return frame("LOOA")


def image(index: int) -> bytes:
    return frame("IMAG", index)


def mode(a: int, b: int = 0) -> bytes:
    """Display mode. Known variants below use specific (a, b) pairs."""
    return frame("MODE", a, b)


def mode_static() -> bytes:
    return frame("MODE", 1)


def mode_flash(rate: int) -> bytes:
    return frame("MODE", 2, rate // 256, rate % 256)


def mode_scroll_left(speed_: int) -> bytes:
    return frame("MODE", 3, speed_)


def mode_scroll_right(speed_: int) -> bytes:
    return frame("MODE", 4, speed_)


def stop_rhythm() -> bytes:
    return frame("STOPR")


def exit_rhythm() -> bytes:
    return frame("SOUT")


def select_lens(which: int) -> bytes:
    """Address one lens: 1 = first, 2 = second."""
    if which == 1:
        return frame("LEDFIRST")
    if which == 2:
        return frame("LEDSECOND")
    raise ValueError("lens must be 1 or 2")


def column(index: int, bitmap: bytes) -> bytes:
    """One display column: [04][index][3 bytes of column bitmap].

    Bit order is not yet confirmed against hardware - see notes/protocol.md.
    """
    if len(bitmap) != 3:
        raise ValueError("column bitmap must be 3 bytes")
    return frame("", *([index] + list(bitmap)))


def columns_from_bitmap(pixels: list[list[int]]) -> list[bytes]:
    """Encode a [rows][cols] 0/1 matrix into per-column frames."""
    if not pixels:
        return []
    rows, cols = len(pixels), len(pixels[0])
    if rows > 24:
        raise ValueError("more than 24 rows does not fit 3 bytes per column")
    out = []
    for c in range(cols):
        bits = 0
        for r in range(rows):
            if pixels[r][c]:
                bits |= 1 << (23 - r)
        out.append(column(c, bits.to_bytes(3, "big")))
    return out

File: packages/core/src/protocol.ts
/**
 * Funky Glasses+ BLE wire format.
 *
 * Recovered from the vendor APK; see notes/protocol.md. Every write is exactly
 * one 16-byte AES-128-ECB block:
 *
 *     [len][opcode ASCII...][args...][zero padding to 16]
 *
 * `len` counts opcode plus arguments, excluding padding.
 */
import { decryptBlock, encryptBlock, expandKey } from './aes.js'

export const SERVICE_UUID = '0000fff0-0000-1000-8000-00805f9b34fb'
export const CHAR_COMMAND = 'd44bc439-abfd-45a2-b575-925416129600'
export const CHAR_NOTIFY = 'd44bc439-abfd-45a2-b575-925416129601'
export const CHAR_BULK_A = 'd44bc439-abfd-45a2-b575-92541612960a'
export const CHAR_BULK_B = 'd44bc439-abfd-45a2-b575-92541612960b'

/** Panchip OTA. Present on the device; writing here can brick it. */
export const SERVICE_OTA = '0000fd00-0000-1000-8000-00805f9b34fb'

export const BLOCK_SIZE = 16
export const NAME_PREFIX = 'GLASSES-'

/** Manufacturer-data marker the vendor app filters on: "TR\0:". */
export const SIGNATURE = new Uint8Array([0x54, 0x52, 0x00, 0x3a])

/** Hardcoded in the vendor's libAES.so, recovered by brute-forcing the image. */
export const KEY = new Uint8Array([
  0x34, 0x52, 0x2a, 0x5b, 0x7a, 0x6e, 0x49, 0x2c,
  0x08, 0x09, 0x0a, 0x9d, 0x8d, 0x2a, 0x23, 0xf8,
])

const ROUND_KEYS = expandKey(KEY)

export const encrypt = (frame: Uint8Array): Uint8Array => encryptBlock(ROUND_KEYS, frame)
export const decrypt = (block: Uint8Array): Uint8Array => decryptBlock(ROUND_KEYS, block)

/** Build a plaintext frame from an ASCII opcode and its arguments. */
export function frame(opcode: string, ...args: number[]): Uint8Array {
  const body = [...opcode].map((c) => c.charCodeAt(0)).concat(args)
  if (body.length > BLOCK_SIZE - 1) {
    throw new Error(`frame body too long: ${body.length}`)
  }
  const out = new Uint8Array(BLOCK_SIZE)
  out[0] = body.length
  out.set(body, 1)
  return out
}

/** The meaningful bytes of a plaintext frame, padding stripped. */
export function body(f: Uint8Array): Uint8Array {
  const n = f[0]
  return n > f.length - 1 ? new Uint8Array(0) : f.subarray(1, 1 + n)
}

const ascii = (b: Uint8Array) => String.fromCharCode(...b)

// --- Command table, transcribed from the vendor app's Agreement.java ---

export const enterDIY = () => frame('SMVEW', 1)
export const enterDIYAlt = () => frame('SMVEW', 3)
export const exitDIY = () => frame('SMVEW', 0)
export const exitDIYSave = () => frame('SMVEW', 2)
export const queryType = () => frame('STYPE')
export const brightness = (level: number) => frame('LIGHT', level)
export const speed = (v: number) => frame('SPEED', v)
export const invert = () => frame('EVERT')
export const animation = (i: number) => frame('ANIM', i)
export const animationLoop = () => frame('LOOA')
export const image = (i: number) => frame('IMAG', i)
export const modeStatic = () => frame('MODE', 1)
export const modeFlash = (rate: number) => frame('MODE', 2, (rate >> 8) & 0xff, rate & 0xff)
export const scrollLeft = (s: number) => frame('MODE', 3, s)
export const scrollRight = (s: number) => frame('MODE', 4, s)
export const stopRhythm = () => frame('STOPR')
export const exitRhythm = () => frame('SOUT')
export const leds = (on: boolean) => (on ? frame('LEDON') : frame('LEDOFF'))
export const flashlight = (on: boolean) => (on ? frame('LIGHTON') : frame('LIGHTOFF'))

/** Address one lens: 1 or 2. Untested against hardware. */
export function lens(which: 1 | 2): Uint8Array {
  return which === 1 ? frame('LEDFIRST') : frame('LEDSECOND')
}

/** One bulk pixel frame: [04][index][3 bytes of column bitmap]. */
export function column(index: number, bitmap: Uint8Array): Uint8Array {
  if (bitmap.length !== 3) throw new Error('column bitmap must be 3 bytes')
  return frame('', index, bitmap[0], bitmap[1], bitmap[2])
}

/** Decode a STYPE reply. Our unit never answers, but the firmware family does. */
export function parseType(f: Uint8Array): { rows: number; cols: number } | null {
  const text = ascii(body(f))
  if (!text.startsWith('STYPE')) return null
  const [r, c] = text.slice(5).split('X')
  const rows = Number(r)
  const cols = Number(c)
  return Number.isFinite(rows) && Number.isFinite(cols) && c !== undefined
    ? { rows, cols }
    : null
}
